Add support for inferring title

jez opened the pull request:
Resolves #64.

The `title` variable in `lib/transform.js` is now populated according to this
logic:

- If the new `--notitle` option is specified:
    - `title = ''`
- If a title was explicitly passed with `--title`:
    - `title` is the user-specified title
- If a TOC exists (as evidenced by `info.hasStart`):
    - `title = lines[info.startIdx + 2]`
    - This line is assumed to always be the TOC title
- Otherwise:
    - `title` is the standard `**Table of Contents**` title

Tests are in `tests/transform-title.js`.


**Commit Summary**

- f6c9e91 - Ignore tags files

- 9508723 - Add clobbertitle option

- f1c6ad8 - Use old title if exists. Also support clobbertitle

- 56bdeaa - Add tests for new title changes

- 4900295 - Clarify README

- 2c69483 - Rename --clobbertitle to --notitle (brevity)

jez commented:
@thlorenz Have you had a chance to look at this? If there's something holding you back I'd love to tweak it.

thlorenz commented on .gitignore:
```diff
@@ -10,6 +10,7 @@ lib-cov
 pids
 logs
 results
+tags
```
Please don't mess with the .gitignore just cause on your system certain extra files are generated.

thlorenz commented on README.md:
```diff
@@ -33,7 +33,7 @@ by github or other sites via a command line flag.
 
 Go into the directory that contains you local git project and type:
-    
+
```
Please fix all white space changes.
I won't merge with them present.

thlorenz commented on test/fixtures/readme-with-custom-title.md:
```diff
@@ -0,0 +1,212 @@
+# dockops [![build status](https://secure.travis-ci.org/thlorenz/dockops.png)](http://travis-ci.org/thlorenz/dockops)
```
Please condense this fixture into a more minimal example.
This way it's hard to see what you're testing since it's so huge.

thlorenz commented on test/transform-html.js:
```diff
@@ -2,7 +2,7 @@
 
 var test = require('tap').test
-  , transform = require('../lib/transform')
+  , transform = require('../lib/transform');
```
Don't mess with my style choices please.

jez replied:
Apologies, I thought this was just a typo.

thlorenz commented on test/transform-html.js:
```diff
@@ -27,7 +27,7 @@ test('\ngiven a file that includes html with header tags and maxHeaderLevel 8',
 
   t.end()
-})
+});
```
Ditto above

thlorenz commented on test/transform-html.js:
```diff
@@ -47,4 +47,4 @@ test('\ngiven a file that includes html with header tags using default maxHeader
   )
   t.end()
-})
+});
```
Ditto above

thlorenz commented on README.md:
```diff
@@ -101,7 +101,9 @@ Running doctoc will insert the toc at that location.
 ### Specifying a custom TOC title
 
-Use the `--title` option to specify a (Markdown-formatted) custom TOC title; e.g., `doctoc --title '**Contents**' .`
+Use the `--title` option to specify a (Markdown-formatted) custom TOC title; e.g., `doctoc --title '**Contents**' .` From then on, you can simply run `doctoc <file>` and doctoc will infer this title from the README.
```
I'd change:
> will infer this title from the README

to:
> will keep the title you specified

thlorenz commented:
In general this looks good to me, although it seems to be a huge effort which you could simply solve via a script in `package.json`, i.e. `"docme": "docme --title my-title README.md"`.

I'm willing to have another look after you fixed all the nits I pointed out.
Thanks.

jez commented:
> In general this looks good to me, although it seems to be a huge effort which you could simply solve via a script in package.json, i.e. "docme": "docme --title my-title README.md".

I thought about this, but decided that baking it into the project was a better option. While it's possible to add it to a script in the `package.json` file, not all projects that have READMEs have a place like this where you can collect short helper scripts.

The next option would be to make it a shell alias, but that doesn't appear automatically across all collaborator's environments. You could also create a shell script and put it in the project folder, but I think it's just easier if this is handled once by doctoc itself.

Just my two cents. I think I addressed all your concerns; if something looks amiss let me know.

jez commented:
@thlorenz I think this might have gotten lost in the noise. Have you had a chance to review it?

thlorenz commented on lib/transform.js:
```diff
@@ -113,15 +113,29 @@ function getLinesToToc (lines, currentToc, info) {
+  else {
+    title = defaultTitle;
+  }
+
```
This block is a little odd.
Couldn't we create a `determineTitle` function and then use early `return` to make this nicer?

```js
function determineTitle(title, notitle, defaultTitle, lines, info) {
  if (notitle) return '';
  if (title) return title;
  return info.hasStart ? lines[..] : defaultTitle; 
}
```

Something like that looks much more readable to me.

thlorenz commented:
Yeah, had slipped my radar. Thanks for the reminder.
So LGTM after that one final nit. Please fix then squash all into one commit and I'll merge.

Thanks.

jez commented:
I added the helper function and squashed.

thlorenz commented:
Thanks merged (ce5b416f87624addd5e97abb5f633830d545698b) and published as 0.14.

thlorenz commented:
Added you as collab so you can merge in the future and help triage issues and answer questions.
Regarding features please stick with the PR (from different branch) workflow so we can discuss before merging to master.
